src/main/scala/is/hail/annotations/ReadRowsRDD.scala

```diff
@@ -145,93 +392,61 @@ class Decoder(in: DataInputStream) {
+final class Encoder(out: OutputBuffer) {
+  def flush() { out.flush() }
 
-  def writeBoolean(b: Boolean) {
-    writeByte(b.toByte)
-  }
-
-  def writeInt(i: Int) {
-    var j = i
-    do {
-      var b = j & 0x7f
-      j >>>= 7
-      if (j != 0)
-        b |= 0x80
-      writeByte(b.toByte)
-    } while (j != 0)
-  }
-
-  def writeLong(l: Long) {
-    var j = l
-    do {
-      var b = j & 0x7f
-      j >>>= 7
-      if (j != 0)
-        b |= 0x80
-      writeByte(b.toByte)
-    } while (j != 0)
-  }
-
-  def writeFloat(f: Float) {
-    out.writeFloat(f)
-  }
-
-  def writeDouble(d: Double) {
-    out.writeDouble(d)
-  }
-
-  def writeBytes(mem: Long, off: Long, n: Int) {
-    var i = 0
-    while (i < n) {
-      out.writeByte(Memory.loadByte(mem + off + i))
-      i += 1
-    }
-  }
+  def writeByte(b: Byte): Unit = out.writeByte(b)
 
   def writeBinary(region: MemoryBuffer, offset: Long) {
     val boff = region.loadAddress(offset)
     val length = region.loadInt(boff)
-    writeInt(length)
-    writeBytes(region.mem, boff + 4, length)
+    out.writeInt(length)
+    out.writeBytes(region.mem, boff + 4, length)
   }
 
   def writeArray(t: TArray, region: MemoryBuffer, aoff: Long) {
     val length = region.loadInt(aoff)
 
     val nMissingBytes = (length + 7) / 8
-    writeInt(length)
-    writeBytes(region.mem, aoff + 4, nMissingBytes)
+    out.writeInt(length)
+    out.writeBytes(region.mem, aoff + 4, nMissingBytes)
 
     val elemsOff = aoff + t.elementsOffset(length)
-    val elemSize = UnsafeUtils.arrayElementSize(t.elementType)
-    var i = 0
-    while (i < length) {
-      if (!region.loadBit(aoff + 4, i)) {
-        val off = elemsOff + i * elemSize
-        t.elementType match {
-          case t2: TStruct => writeStruct(t2, region, off)
-          case t2: TArray => writeArray(t2, region, region.loadAddress(off))
-          case TBoolean => writeBoolean(region.loadByte(off) != 0)
-          case TInt32 => writeInt(region.loadInt(off))
-          case TInt64 => writeLong(region.loadLong(off))
-          case TFloat32 => writeFloat(region.loadFloat(off))
-          case TFloat64 => writeDouble(region.loadDouble(off))
-          case TBinary => writeBinary(region, off)
+    val elemSize = t.elementByteSize
+    if (t.elementType == TInt32) { // fast case
+      var i = 0
+      while (i < length) {
+        if (t.isElementDefined(region, aoff, i)) {
+          val off = elemsOff + i * elemSize
+          out.writeInt(region.loadInt(off))
         }
+        i += 1
       }
+    } else {
+      var i = 0
+      while (i < length) {
+        if (t.isElementDefined(region, aoff, i)) {
+          val off = elemsOff + i * elemSize
+          t.elementType match {
+            case t2: TStruct => writeStruct(t2, region, off)
+            case t2: TArray => writeArray(t2, region, region.loadAddress(off))
+            case TBoolean => out.writeBoolean(region.loadByte(off) != 0)
+            case TInt32 => out.writeInt(region.loadInt(off))
```

Comment:
`assert(false)`?

Reply:
Where?  I don't understand this.

Reply:
The `TInt32` case is unreachable because of your fast path, so I'm inclined to assert the fact that the code is unreachable.

Reply:
Ah, right!